gui/journal.lua

```diff
@@ -0,0 +1,797 @@
+-- Fort journal with a multi-line text editor
+
+local gui = require 'gui'
+local widgets = require('gui.widgets')
+
+local CLIPBOARD_MODE = {LOCAL = 1, LINE = 2}
+
+TextEditor = defclass(TextEditor, widgets.Widget)
+
+TextEditor.ATTRS{
+    text = '',
+    text_pen = COLOR_LIGHTCYAN,
+    ignore_keys = {'STRING_A096'},
+    select_pen = COLOR_CYAN,
+    on_change = DEFAULT_NIL,
+    debug = false
+}
+
+-- similar to string:wrap, but do not skip any spaces and new lines characters
+-- it returns table of lines on the output, instead of string. such table items
+-- contacted will always generate exactly the same text like provided to the fun
+function string_strict_wrap(text, width)
```

Comment:
let me see if I can offer this as a library function.

I think the current `string:wrap` is wrong to condense spaces. I can make that change at the library level. Then, we can offer a variant of the function that returns the table of strings instead of a pre-concatenated string

Reply:
done in https://github.com/DFHack/dfhack/pull/4758

Reply:
It almost works, but it skip new lines character, which I need in my logic.
the information is already here, as table elements, so I added the newline characters by myself,
plz consider if it would be useful in original script to keep newline characters controlled by a flag or smth.

Reply:
I realized the current version of `text:wrap` is not enough for the multiline editor.
As it skip new characters, I can not recognize when the text has been wrapped, from where text has been break by new line. Its the source of errors with the cursor you found:
https://github.com/DFHack/scripts/pull/1208#issuecomment-2197565354